Description: Compatibility fixes for Python 3.15
 - Provide a compatibility implementation of PyWeakref_GetObject and
   PyWeakref_GET_OBJECT using PyWeakref_GetRef.
 - Fix undefined _Py_tss_interp in atexit.pyx by defining Py_BUILD_CORE_MODULE
   and using PyInterpreterState_Get().
Bug-Sage: https://github.com/sagemath/sage/issues/42891
Author: Maximiliano Curia <maxy@debian.org>

Index: sagemath/src/meson.build
===================================================================
--- sagemath.orig/src/meson.build	2026-10-07 18:13:58.778175137 +0200
+++ sagemath/src/meson.build	2026-10-07 18:13:58.777823512 +0200
@@ -292,6 +292,10 @@
   '-X preliminary_late_includes_cy28=True',
   language: 'cython',
 )
+add_project_arguments(
+  '-include', meson.current_source_dir() / 'sage' / 'cpython' / 'compat.h',
+  language: ['c', 'cpp'],
+)
 
 inc_cpython = include_directories('sage/cpython')
 inc_rings = include_directories('sage/rings')
Index: sagemath/src/sage/cpython/compat.h
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ sagemath/src/sage/cpython/compat.h	2026-10-07 18:13:58.778030595 +0200
@@ -0,0 +1,38 @@
+#ifndef SAGE_CPYTHON_COMPAT_H
+#define SAGE_CPYTHON_COMPAT_H
+
+#include <patchlevel.h>
+
+#if PY_VERSION_HEX >= 0x030f0000
+#include <Python.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+static inline PyObject* PyWeakref_GetObject(PyObject *ref)
+{
+    PyObject *obj = NULL;
+    if (ref == NULL) {
+        PyErr_BadInternalCall();
+        return NULL;
+    }
+    if (PyWeakref_GetRef(ref, &obj) < 0) {
+        return NULL;
+    }
+    if (obj == NULL) {
+        return Py_None;
+    }
+    Py_DECREF(obj);
+    return obj;
+}
+
+#define PyWeakref_GET_OBJECT(ref) PyWeakref_GetObject((PyObject *)(ref))
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif
+
+#endif
Index: sagemath/src/sage/cpython/meson.build
===================================================================
--- sagemath.orig/src/sage/cpython/meson.build	2026-10-07 18:13:58.778175137 +0200
+++ sagemath/src/sage/cpython/meson.build	2026-10-07 18:13:58.778169166 +0200
@@ -3,6 +3,7 @@
   'all.py',
   'atexit.pyx',
   'builtin_types.pyx',
+  'compat.h',
   'cython_metaclass.h',
   'cython_metaclass.pxd',
   'cython_metaclass.pyx',
Index: sagemath/src/sage/cpython/atexit.pyx
===================================================================
--- sagemath.orig/src/sage/cpython/atexit.pyx	2026-09-23 22:00:16.771863106 +0200
+++ sagemath/src/sage/cpython/atexit.pyx	2026-10-07 18:14:27.772748463 +0200
@@ -152,6 +152,9 @@
     #ifndef Py_BUILD_CORE
     #define Py_BUILD_CORE
     #endif
+    #ifndef Py_BUILD_CORE_MODULE
+    #define Py_BUILD_CORE_MODULE 1
+    #endif
     #undef _PyGC_FINALIZED
     #include "internal/pycore_interp.h"
     #include "internal/pycore_pystate.h"
@@ -168,7 +171,7 @@
     // Note: In Python 3.14 the atexit_state struct changed - callbacks is now a PyObject* (PyList)
     
     static PyObject* get_atexit_callbacks_list(PyObject *self) {
-        PyInterpreterState *interp = _PyInterpreterState_GET();
+        PyInterpreterState *interp = PyInterpreterState_Get();
         // Access the callbacks list directly from the interpreter state
         // We return a new reference because Cython expects an owned reference
         PyObject *callbacks = interp->atexit.callbacks;
@@ -184,7 +187,7 @@
     #else
     // Python < 3.14: atexit uses C array
     static atexit_callback_struct** get_atexit_callbacks_array(PyObject *self) {
-        PyInterpreterState *interp = _PyInterpreterState_GET();
+        PyInterpreterState *interp = PyInterpreterState_Get();
         struct atexit_state state = interp->atexit;
         // Cast from atexit_callback** to our struct type
         return (atexit_callback_struct**)state.callbacks;
